From 9512d78a72316932e01ead2957784fab20f3e575 Mon Sep 17 00:00:00 2001
From: Niklas Hauser <niklas.hauser@rwth-aachen.de>
Date: Mon, 12 Oct 2026 18:30:31 +0200
Subject: [PATCH 3/6] Compare ASCII keys of case insensitive dictionaries
 without case folding

QString::compare() with Qt::CaseInsensitive case folds both strings into
new strings on every call. The global class dictionary is sorted case
insensitive by default and is searched for nearly every word of the
documentation, so the lookups spent most of their time allocating and
folding strings.

Compare ASCII characters in place and only fall back to
QString::compare() when a non ASCII character is found. The result is
identical.
---
 src/stringmap.h | 52 +++++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 50 insertions(+), 2 deletions(-)

diff --git a/src/stringmap.h b/src/stringmap.h
index f27f27b8..33f88e3e 100644
--- a/src/stringmap.h
+++ b/src/stringmap.h
@@ -61,15 +61,63 @@ class StringCompare {
          : m_sortCase_enum(sortCase), m_uniqueCase(uniqueCase) { }
 
       bool operator()(const QString &a, const QString &b) const {
-         int retval = a.compare(b, m_sortCase_enum);
+         int retval;
 
-         if (retval == 0 && m_uniqueCase && m_sortCase_enum == Qt::CaseInsensitive) {
+         if (m_sortCase_enum == Qt::CaseInsensitive) {
+            retval = compareNoCase(a, b);
+
+            if (retval == 0 && m_uniqueCase) {
+               retval = a.compare(b, Qt::CaseSensitive);
+            }
+
+         } else {
             retval = a.compare(b, Qt::CaseSensitive);
          }
 
          return retval < 0;
       }
 
+      // same result as a.compare(b, Qt::CaseInsensitive), which case folds both strings into new
+      // strings for every call, ASCII text is compared in place since this is used for every lookup
+      static int compareNoCase(const QString &a, const QString &b) {
+         auto iterA = a.storage_begin();
+         auto iterB = b.storage_begin();
+
+         const auto endA = a.storage_end();
+         const auto endB = b.storage_end();
+
+         while (iterA != endA && iterB != endB) {
+            unsigned char chA = *iterA;
+            unsigned char chB = *iterB;
+
+            if (chA >= 0x80 || chB >= 0x80) {
+               return a.compare(b, Qt::CaseInsensitive);
+            }
+
+            if (chA >= 'A' && chA <= 'Z') {
+               chA += 'a' - 'A';
+            }
+
+            if (chB >= 'A' && chB <= 'Z') {
+               chB += 'a' - 'A';
+            }
+
+            if (chA != chB) {
+               return chA < chB ? -1 : 1;
+            }
+
+            ++iterA;
+            ++iterB;
+         }
+
+         if (iterA == endA && iterB == endB) {
+            return 0;
+         }
+
+         // remaining text can contain non ASCII characters which fold to nothing or to ASCII
+         return a.compare(b, Qt::CaseInsensitive);
+      }
+
    private:
       Qt::CaseSensitivity m_sortCase_enum;
       bool m_uniqueCase;
